sources/api/prairiedog/src/bootconfig.rs

```diff
@@ -0,0 +1,194 @@
+use argh::FromArgs;
+use bytes::BufMut;
+use model::BootSettings;
+use prairiedog::de::from_str;
+use prairiedog::ser::to_string;
+use snafu::{OptionExt, ResultExt};
+use std::convert::TryInto;
+use std::io;
+use std::mem::size_of;
+use std::path::Path;
+use tokio::fs;
+
+// Boot config/initrd related consts
+const BASE_INITRD_SIZE: usize = 0;
+const BOOTCONFIG_MAGIC: &str = "#BOOTCONFIG\n";
+// 1 << 2 = 4
+const BOOTCONFIG_ALIGN: usize = 0b100;
+const BOOTCONFIG_ALIGN_MASK: usize = 0b11;
+const BOOTCONFIG_INITRD_PATH: &str = "/var/lib/bottlerocket/bootconfig.data";
+const PROC_BOOTCONFIG: &str = "/proc/bootconfig";
+
+#[derive(FromArgs, PartialEq, Debug)]
+#[argh(subcommand, name = "generate-boot-config")]
+/// Generate boot configuration from settings
+pub(crate) struct GenerateBootConfigArgs {}
+
+#[derive(FromArgs, PartialEq, Debug)]
+#[argh(subcommand, name = "generate-boot-settings")]
+/// Generate boot settings from existing boot configuration
+pub(crate) struct GenerateBootSettingsArgs {}
+
+/// Queries Bottlerocket boot settings and generates initrd image file with boot config as the only data
+pub(crate) async fn generate_boot_config<P>(socket_path: P) -> Result<()>
+where
+    P: AsRef<Path>,
+{
+    let boot_settings = get_boot_config_settings(socket_path).await?;
+    info!("Generating initrd boot config from boot settings");
+    trace!("Boot settings: {:?}", boot_settings);
+    let bootconfig = to_string(&boot_settings).context(error::SerializeBootConfigSnafu)?;
+    trace!("Serializied boot config string: {}", bootconfig);
+    let initrd = generate_initrd(&bootconfig.into_bytes())?;
+    trace!("Writing initrd image file: {:?}", initrd);
+    tokio::fs::write(BOOTCONFIG_INITRD_PATH, &initrd)
+        .await
+        .context(error::WriteInitrdSnafu)?;
+    Ok(())
+}
+
+async fn get_boot_config_settings<P>(socket_path: P) -> Result<BootSettings>
+where
+    P: AsRef<Path>,
+{
+    let boot_settings = schnauzer::get_settings(socket_path)
+        .await
+        .context(error::RetrieveSettingsSnafu)?
+        .settings
+        .and_then(|s| s.boot)
+        .context(error::MissingSettingsSnafu)?;
+    Ok(boot_settings)
+}
+
+fn generate_initrd(bootconfig: &[u8]) -> Result<Vec<u8>> {
+    let mut bootconfig_size = bootconfig.len();
+    // Calculate checksum of the boot config file
+    let mut checksum: u32 = 0;
+    for byte in bootconfig {
+        checksum = checksum
+            .checked_add(u32::from(*byte))
+            .context(error::AddU32OverflowSnafu)?;
+    }
+    // Bottlerocket does not use an initrd, so the base initrd image size is 0
+    let total_size = BASE_INITRD_SIZE
+        + bootconfig_size
+        + size_of::<u32>() * 2
+        + BOOTCONFIG_MAGIC.as_bytes().len();
+    let mut initrd = bootconfig.to_owned();
+    // initrd image file needs to be 4-byte aligned, so we add padding as necessary.
```

Comment:
An extra comment or two here about the math would be helpful! :) (since we don't do bitwise stuff often)